feat(device-open-mac): added non exclusive device open

Pen7art opened the pull request:
Added a way to open devices in non-exclusive mode on the mac platform. 
The user still has to give input permissions so I don't think this can create security issues.
The code builds and works on my mac device but I'm not 100% confident that the build process is correct since this is my first time working with something like this.

Julusian reviewed (commented):
I've added some comments, but I am wondering if this is the right approach?

My concern is that I have an application which has node-hid as a dependency of 3+ other libraries. If one of those sets this, that could cause the other libraries to make bad assumptions because they are not in the exclusivity mode they expect.

So I feel like this would be better to have as an optional argument to the function when opening a device. eg:
```
// With it being defined:
const dev = new HID('/device-path/here', { exclusive: false })
const dev = new HID(0x01, 0x02, { exclusive: false })

// Backwards compatible to before:
const dev = new HID('/device-path/here', {  })
const dev = new HID('/device-path/here')
```
That way we avoid libraries fighting over what the value of this property should be.

Looking at the implementation of `hid_darwin_set_open_exclusive`, there won't be any noticable cost to calling it before every open of a hid device

ert78gb commented:
I agree with Julusian, the "non exclusive" flag should be optional parameter of each `HID` object.

Julusian commented:
I realise there is no example of how to do that optional options object in this repostitory, here is one from another of mine, that is hopefully a good enough reference to adapt:
https://github.com/Julusian/node-jpeg-turbo/blob/683936714140efd9386991836c7ba62a7490493d/src/decompress.cc#L133-L153
And the method signature, to help you figure out why the code is doing some things: https://github.com/Julusian/node-jpeg-turbo/blob/683936714140efd9386991836c7ba62a7490493d/module.d.ts#L60-L65

Pen7art commented:
@Julusian I've updated the PR to include your suggestions. Mainly I added the nonExclusive  optional boolean parameter to the open function. I used nonExclusive as a name because the default behaviour is to open in exclusive mode so i thought nonExclusive made more sense. The only things I could not figure out is how to improve the "../hidapi/mac/hidapi_darwin.h" import. Let me know if you have some suggestions about that.

ert78gb commented on nodehid.d.ts:
```diff
@@ -21,8 +21,8 @@ export interface Device {
 export class HID extends EventEmitter {
-    constructor(path: string)
-    constructor(vid: number, pid: number)
+    constructor(path: string, { nonExclusive }: { nonExclusive?: boolean })
```
I think the whole `options` argument should be optional.

Julusian commented on src/HID.cc:
```diff
@@ -45,13 +49,38 @@ HID::HID(const Napi::CallbackInfo &info)
 
-  if (info.Length() == 1)
+#if defined(__APPLE__)
+  if (argsLength > 1)
+  {
+    // We check if we have an optional param
+    if (info[argsLength - 1].IsObject())
+    {
+      argsLength -= 1;
+      Napi::Object options = info[argsLength].As<Napi::Object>();
+
+      Napi::Value isNonExclusiveMode = options.Get("nonExclusive");
+      if (!isNonExclusiveMode.IsBoolean())
+      {
+        Napi::TypeError::New(env, "nonExclusive flag should be a boolean").ThrowAsJavaScriptException();
+        return;
+      }
+      if (isNonExclusiveMode.As<Napi::Boolean>().Value())
+      {
+        hid_darwin_set_open_exclusive(0);
+      }
+    }
+  }
```
```suggestion
  bool isNonExclusiveBool = false
  if (argsLength > 1)
  {
    // We check if we have an optional param
    if (info[argsLength - 1].IsObject())
    {
      argsLength -= 1;
      Napi::Object options = info[argsLength].As<Napi::Object>();

      Napi::Value isNonExclusiveMode = options.Get("nonExclusive");
      if (!isNonExclusiveMode.IsBoolean())
      {
        Napi::TypeError::New(env, "nonExclusive flag should be a boolean").ThrowAsJavaScriptException();
        return;
      }
      isNonExclusiveBool = isNonExclusiveMode.As<Napi::Boolean>().Value();
    }
  }
  
  hid_darwin_set_open_exclusive(isNonExclusiveBool ? 0 : 1);
```

This needs to be able to both enable and disable this flag. And it should do so even when the options object is not defined

Julusian commented on src/HIDAsync.cc:
```diff
@@ -216,7 +222,27 @@ Napi::Value HIDAsync::Create(const Napi::CallbackInfo &info)
   ContextState *context = (ContextState *)data;
 
-  if (info.Length() == 1)
+#if defined(__APPLE__)
+  if (argsLength > 1)
+  {
+    // We check if we have an optional param
+    if (info[argsLength - 1].IsObject())
+    {
+      argsLength -= 1;
+      Napi::Object options = info[argsLength].As<Napi::Object>();
+
+      Napi::Value isNonExclusiveMode = options.Get("nonExclusive");
+      if (!isNonExclusiveMode.IsNumber())
+      {
+        Napi::TypeError::New(env, "nonExclusive flag should be a boolean").ThrowAsJavaScriptException();
+        return env.Null();
+      }
+      hid_darwin_set_open_exclusive(isNonExclusiveMode.As<Napi::Number>().Int32Value());
+    }
+  }
+#endif
```
This block is a little different to the one in HID.cc. It should be the same

Pen7art commented:
I've implemented your suggestions let me know if there is anything else you guys would like to change

ert78gb commented:
The typescript part is ok for me. My C knowledge is not on an approver level.

Thx for your effort

Pen7art commented:
@todbot do you think we can merge this?

TheCrypt0 commented:
This would be super useful for me too! Can it be merged? 

@todbot @ert78gb @Julusian
